File: ios/EpycCore/Sources/EpycCore/Navigation.swift
import Foundation

/// Where the state-driven router should send a player, ported from
/// `src/providers/game-navigation-controller/game-navigation-controller.ts`.
/// Raw values match the golden vectors' `destination` strings.
public enum Destination: String, Sendable, Equatable {
    case waitingRoom = "WAITING_ROOM"
    case draw = "DRAW"
    case guess = "GUESS"
    case waitTurn = "WAIT_TURN"
    case waitGameToEnd = "WAIT_GAME_TO_END"
    case gameResults = "GAME_RESULTS"
}

public enum GameNavigation {

    /// Pure screen-selection decision. The view layer adds parameters (word /
    /// drawingRef) once the destination is known.
    public static func destination(state: GameState,
                                   threads: [GameThread],
                                   usersOrder: [String],
                                   userId: String) -> Destination {
        if state == .created { return .waitingRoom }
        guard state == .started, usersOrder.contains(userId) else { return .gameResults }
        if TurnEngine.isGameDone(threads: threads) { return .gameResults }

        let next = TurnEngine.getNextAtom(threads: threads, usersOrder: usersOrder, userId: userId)
        if next.allAtomsDone { return .waitGameToEnd }
        if !next.readyToPlay { return .waitTurn }
        if let addr = next.address {
            let atom = threads[addr.threadIndex].gameAtoms[addr.atomIndex]
            return atom.type == .drawing ? .draw : .guess
        }
        return .gameResults
    }

    /// The word a DRAW atom should show: the thread word for atom 0, otherwise the
    /// previous atom's guess. (Ported from getNavigationTargetFromGameState.)
    public static func drawWord(threads: [GameThread], address: AtomAddress) -> String? {
        let thread = threads[address.threadIndex]
        if address.atomIndex == 0 { return thread.word }
        return thread.gameAtoms[address.atomIndex - 1].guess
    }

    /// The drawing a GUESS atom should replay: the previous atom's drawingRef.
    public static func guessDrawingRef(threads: [GameThread], address: AtomAddress) -> String? {
        guard address.atomIndex > 0 else { return nil }
        return threads[address.threadIndex].gameAtoms[address.atomIndex - 1].drawingRef
    }
}

File: android/core/src/main/kotlin/com/negfeed/epyc/core/Navigation.kt
package com.negfeed.epyc.core

/**
 * Where the state-driven router should send a player, ported from
 * `src/providers/game-navigation-controller/game-navigation-controller.ts`.
 * `value` matches the golden vectors' `destination` strings.
 */
enum class Destination(val value: String) {
    WAITING_ROOM("WAITING_ROOM"),
    DRAW("DRAW"),
    GUESS("GUESS"),
    WAIT_TURN("WAIT_TURN"),
    WAIT_GAME_TO_END("WAIT_GAME_TO_END"),
    GAME_RESULTS("GAME_RESULTS"),
}

object GameNavigation {

    fun destination(
        state: GameState,
        threads: List<GameThread>,
        usersOrder: List<String>,
        userId: String,
    ): Destination {
        if (state == GameState.CREATED) return Destination.WAITING_ROOM
        if (state != GameState.STARTED || !usersOrder.contains(userId)) return Destination.GAME_RESULTS
        if (TurnEngine.isGameDone(threads)) return Destination.GAME_RESULTS

        val next = TurnEngine.getNextAtom(threads, usersOrder, userId)
        if (next.allAtomsDone) return Destination.WAIT_GAME_TO_END
        if (!next.readyToPlay) return Destination.WAIT_TURN
        val addr = next.address ?: return Destination.GAME_RESULTS
        val atom = threads[addr.threadIndex].gameAtoms[addr.atomIndex]
        return if (atom.type == GameAtomType.DRAWING) Destination.DRAW else Destination.GUESS
    }

    /** Word a DRAW atom shows: thread word for atom 0, else previous atom's guess. */
    fun drawWord(threads: List<GameThread>, address: AtomAddress): String? {
        val thread = threads[address.threadIndex]
        return if (address.atomIndex == 0) thread.word
        else thread.gameAtoms[address.atomIndex - 1].guess
    }

    /** Drawing a GUESS atom replays: previous atom's drawingRef. */
    fun guessDrawingRef(threads: List<GameThread>, address: AtomAddress): String? {
        if (address.atomIndex <= 0) return null
        return threads[address.threadIndex].gameAtoms[address.atomIndex - 1].drawingRef
    }
}
